tools/access_replay.c

```diff
@@ -0,0 +1,629 @@
+	{
+		guint64 size = parse_ul(parts[SIZE]);
+		if (size > memory_chunk_size)
+		{
+			g_warning("unable to replay: chunk size to samll!%lu vs %lu", size, memory_chunk_size);
+		}
+		else
+		{
+			ret = j_backend_kv_put(kv_backend, batch, parts[NAME], memory_chunk, size);
+		}
+	}
+	else if (strcmp(op, "delete") == 0)
+	{
+		ret = j_backend_kv_delete(kv_backend, batch, parts[NAME]);
+	}
+	else if (strcmp(op, "get") == 0)
+	{
+		guint32 size;
+		gpointer value;
+		ret = j_backend_kv_get(kv_backend, batch, parts[NAME], &value, &size);
+	}
+	else if (strcmp(op, "get_all") == 0)
+	{
+		if (batch != NULL)
+		{
+			g_warning("start iterating with remaining batch or iteration");
+		}
+		ret = j_backend_kv_get_all(kv_backend, parts[NAMESPACE], &batch);
+	}
+	else if (strcmp(op, "get_by_prefix") == 0)
+	{
+		if (batch != NULL)
+		{
+			g_warning("start iterating with remaining batch or iteration");
+		}
+		ret = j_backend_kv_get_by_prefix(kv_backend, parts[NAMESPACE], parts[NAME], &batch);
+	}
+	else if (strcmp(op, "iterate") == 0)
+	{
+		const char* name;
+		gconstpointer value;
+		guint32 len;
+		ret = j_backend_kv_iterate(kv_backend, batch, &name, &value, &len);
+		if (!ret)
+		{
+			batch = NULL;
+			ret = TRUE;
+		}
+	}
+	else if (strcmp(op, "init") == 0)
+	{
+		ret = TRUE;
+	}
+	else if (strcmp(op, "fini") == 0)
+	{
+		ret = TRUE;
+	}
+	else if (strcmp(op, "batch_execute") == 0)
+	{
+		ret = j_backend_kv_batch_execute(kv_backend, batch);
+	}
+	else
+	{
+		g_warning("unkown operation: '%s'", op);
+	}
+	return ret;
+}
+struct JSqlBatch
+{
+	const gchar* namespace;
+	JSemantics* semantics;
+	gboolean open;
+	gboolean aborted;
+};
+static gboolean
+replay_db(const char* parts[ROW_LEN])
+{
+	gboolean ret = FALSE;
+	static gpointer batch = NULL;
+	static gpointer itr = NULL;
+	static gchar* namespace = NULL;
+	/// \TODO how to determine semantics used?
+	static JSemantics* semantics = NULL;
+	GError* error = NULL;
+	const char* op = parts[OPERATION];
+	static GArray* bsons = NULL;
+	if (bsons == NULL)
+	{
+		bsons = g_array_sized_new(FALSE, TRUE, sizeof(bson_t*), 0);
+	}
+	if (semantics == NULL)
+	{
+		semantics = j_semantics_new(J_SEMANTICS_TEMPLATE_DEFAULT);
+	}
+
+	if (strcmp(op, "batch_start") == 0)
+	{
+		if (batch != NULL)
+		{
+			g_warning("starting a new batch, but old one is not finished");
+		}
+		namespace = strdup(parts[NAMESPACE]);
+		ret = j_backend_db_batch_start(db_backend, namespace, semantics, &batch, &error);
+	}
+	else if (strcmp(op, "schema_create") == 0)
+	{
+		bson_t* schema = bson_new_from_json((const uint8_t*)parts[JSON], -1, NULL);
+		ret = j_backend_db_schema_create(db_backend, batch, parts[NAME], schema, &error);
+		g_array_append_val(bsons, schema);
+	}
+	else if (strcmp(op, "schema_get") == 0)
+	{
+		bson_t* schema = bson_new();
+		ret = j_backend_db_schema_get(db_backend, batch, parts[NAME], schema, &error);
+		g_array_append_val(bsons, schema);
+		// schema do not exists
+		if (error && error->code == 8)
```

Comment:
Normal behavior is to check if i schema exists before creating a new one, therefore a `schema does not exist` is no error to panic